### PR TITLE
feat(llvm): Make NaN canonicalization configurable

### PR DESCRIPTION
**Pull request by jubianchi**
compiler-llvm now uses the experimental.constrained intrinsics to ensure
correct behavior on FP operations when full-canonicalization is
disabled.
    
This patch requires TheDan64/inkwell#247

# Review

- [ ] Add a short description of the the change to the CHANGELOG.md file

### CODE REVIEW

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

File: lib/c-api/src/wasm_c_api/engine.rs
Comment (MarkMcCaskey) [inline lib/c-api/src/wasm_c_api/engine.rs:102]: If this is for the spec-test compliant part of nan canonicalization and not the "canonicalize everywhere" thing we also offer, I'd make it more explicit that this will make the Wasm VM behave in a non-spec compliant way. Probably in the name of the function or something, include the word "unsafe", like what React does.

**Comment by jubianchi** [comment]
bors try

File: lib/c-api/src/wasm_c_api/engine.rs
Comment (syrusakbary) [inline lib/c-api/src/wasm_c_api/engine.rs:308]: It might be a good idea to add a second variable (enable: true/false), in case someone want to overwrite the default value (and that default value is true)

File: lib/compiler-singlepass/src/config.rs
Comment (syrusakbary) [inline lib/compiler-singlepass/src/config.rs:22]: I think since singlepass is mainly directed for blockchains, which need deterministic results by default, this value should be true by default.

File: lib/compiler/src/compiler.rs
Comment (syrusakbary) [inline lib/compiler/src/compiler.rs:43]: Receive another variable, and make the function not have a default implementation.

```suggestion
    fn enable_nan_canonicalization(&mut self, enable: bool);
```

File: tests/ignores.txt
Comment (syrusakbary) [inline tests/ignores.txt:39]: TODO: Don't merge this. When testing, we should enable the canonical flag.

File: lib/compiler/src/compiler.rs
Comment (nlewycky) [reply lib/compiler/src/compiler.rs:43]: If it takes a bool to enable or disable, maybe `set_nan_canonicalization` would be preferred? Or `set_canonicalize_all_nans`?

Comment (syrusakbary) [reply lib/compiler/src/compiler.rs:43]: Yeah, perhaps naming can be better in that case!

Comment (syrusakbary) [reply lib/compiler/src/compiler.rs:43]: For the Wasm features we simply put the name directly (without set/enable).
https://github.com/wasmerio/wasmer/blob/master/lib/wasmer-types/src/features.rs#L107-L110

Perhaps
`canonicalize_nans(&mut self, enable: bool);` ?

File: lib/c-api/src/wasm_c_api/engine.rs
Comment (Hywan) [reply lib/c-api/src/wasm_c_api/engine.rs:308]: This function must move into the `unstable/engine.rs` module for the moment.
Also, I propose the following API: `wasm_config_nan_canonicalization(config;: &mut wasm_config_t, enable: bool)`. Thoughts?

Finally, this function should be added inside the `build.rs` script, inside the `exclude_items_from_wasm_c_api` function, so that it's not present in the `wasmer.h*` header files.

File: lib/compiler-llvm/src/config.rs
Comment (syrusakbary) [inline lib/compiler-llvm/src/config.rs:56]: Lets have it false by default

File: tests/wast/wasmer/nan-canonicalization.wast
Comment (syrusakbary) [inline tests/wast/wasmer/nan-canonicalization.wast:194]: Will this change work with the other compilers??

Comment (syrusakbary) [reply tests/wast/wasmer/nan-canonicalization.wast:194]: Ok, I think we have to replace all this numbers with:
* `0x7ff8000000000001` => `nan`
* `0x7ff8000000000000` => `nan:canonical`

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

File: lib/compiler-llvm/src/translator/code.rs
Comment (nlewycky) [inline lib/compiler-llvm/src/translator/code.rs:3266]: Using `ExtraInfo` we used to be able to detect when the NaN is certainly already canonical and not re-canonicalize it here

File: lib/compiler-llvm/src/translator/code.rs
Comment (nlewycky) [inline lib/compiler-llvm/src/translator/code.rs:3275]: It'd be nice if this was named to say what the rounding mode is. Similarly for exception.

File: lib/compiler-llvm/src/translator/code.rs
Comment (nlewycky) [inline lib/compiler-llvm/src/translator/code.rs:3955]: This much be replaced with a call to `llvm.experimental.constrained.fcmp`:
> If any FP operation in a function is constrained then they all must be constrained. This is required for correct LLVM IR.
-- https://llvm.org/docs/LangRef.html#constrained-floating-point-intrinsics

Please audit for any other non-constrainted floating point operations.

File: lib/compiler-llvm/src/translator/intrinsics.rs
Comment (nlewycky) [inline lib/compiler-llvm/src/translator/intrinsics.rs:111]: > All function calls done in a function that uses constrained floating point intrinsics must have the `strictfp` attribute.
> All function definitions that use constrained floating point intrinsics must have the `strictfp` attribute.
-- https://llvm.org/docs/LangRef.html#constrained-floating-point-intrinsics

Please add the `strictfp` attribute and place it in all places where it's required, which includes the trampolines.

**Review by nlewycky — COMMENTED** [review]
I don't understand the changes to llvm at all. The constrained FP functions don't appear to have any documented difference in how they handle NaNs than the non-constainted operations. They only let you specify rounding mode and floating-point-exception handling. Also, this removed all the work done to track the state of our FP values to know whether they may contain NaNs and therefore don't need to be canonicalized, instead always inserting unnecessary canonicalization operations. What am I not seeing?

File: lib/c-api/build.rs
Comment (Hywan) [inline lib/c-api/build.rs:479]:
```suggestion
        .exclude_item("wasm_config_canonicalize_nans")
```

File: lib/compiler-cranelift/src/config.rs
Comment (Hywan) [inline lib/compiler-cranelift/src/config.rs:191]: Why not an API like `enable_nan_canonicalization(&mut self, enable: bool)`, just like we have for `enable_simd` and probably others?

Comment (jubianchi) [reply lib/compiler-cranelift/src/config.rs:191]: my main concern here is consistency.

On `CompilerConfig` all the other things are enabled through `enable_<something>(&mut self)` functions.

If I decide to add a `enable: bool` parameter to `enable_nan_canonicalization` then I should add the same to all the other `enable_<something>` which is (strictly speaking) a breaking change to a public API that is most likely used a lot.

This is why I introduce a new way to switch this feature on or off. But I agree, the name can be better, something like `toggle_nan_canonicalization(&mut self, enable: bool)` would be better. and then, we can add `enable_<something>(&mut self, enable: bool)`.

WDYT?

File: tests/compilers/wast.rs
Comment (syrusakbary) [inline tests/compilers/wast.rs:55]: I think all compilers should behave the same way if a feature X is enabled or disabled.
So I don't think is a good idea to change the features depending on the compiler used (other than support or not support something)

File: lib/compiler-llvm/Cargo.toml
Comment (syrusakbary) [inline lib/compiler-llvm/Cargo.toml:29]: Let's make sure a version is published to creates.io (we can also create `wasmer-inkwell`) before merging

File: lib/compiler-cranelift/src/config.rs
Comment (Hywan) [reply lib/compiler-cranelift/src/config.rs:191]: Let's keep things as they are for the moment.

**Comment by syrusakbary** [comment]
bors try

File: lib/compiler-llvm/src/abi/aarch64_systemv.rs
Comment (MarkMcCaskey) [inline lib/compiler-llvm/src/abi/aarch64_systemv.rs:88]: The semantics here have changed from returning an error value if any of these things fail to just panicking. You can probably preserve the collect to `Result<Vec<_>, _>` behavior if you want with something like

```param_types.map(|v| v.and_then(|v| Ok(v.into()))).collect::<Result<Vec<BasicMetadataTypeEnum>, _>()?```

File: lib/compiler-llvm/src/translator/code.rs
Comment (MarkMcCaskey) [inline lib/compiler-llvm/src/translator/code.rs:2232]: ⛳ You can also express this as:

```Rust
params.iter().copied().map(Into::into).collect::<Vec<BasicMetadataValueEnum>().as_slice()
```

if you prefer

Comment (MarkMcCaskey) [reply lib/compiler-llvm/src/translator/code.rs:2232]: You  can also make `From` impls that accept `&T` as well so you don't have to worry about it being `T` or `&T`, then you can remove the `copied` call here

File: lib/compiler-llvm/src/translator/code.rs
Comment (MarkMcCaskey) [inline lib/compiler-llvm/src/translator/code.rs:4380]: Does it make sense to change the return value of `pop1` etc.? Do we ever use the non-`into`d version?

File: lib/compiler-llvm/src/translator/intrinsics.rs
Comment (MarkMcCaskey) [inline lib/compiler-llvm/src/translator/intrinsics.rs:295]: We could also cache the most common ones here like we did before

File: lib/compiler-llvm/src/translator/code.rs
Comment (jubianchi) [reply lib/compiler-llvm/src/translator/code.rs:4380]: Actually, we can't. Doing this will just make us move the `into()` calls to other places.

For now, I think we can keep everything as is.

**Review by Hywan — APPROVED** [review]
Good work!

**Pull request merged**
